src/data/list/pi.lean

```diff
@@ -0,0 +1,45 @@
+/-
+Copyright (c) 2023 Yuyang Zhao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuyang Zhao
+-/
+import tactic.ext
+import tactic.split_ifs
+
+/-!
+# The cartesian product of lists
+-/
+
+namespace list
+variables {ι : Type*} [decidable_eq ι] {α : ι → Sort*}
+
+namespace pi
+
+/-- Given `α : ι → Sort*`, `pi.nil α` is the trivial dependent function out of the empty list. -/
+def nil {ι : Type*} (α : ι → Sort*) : (Π i ∈ ([] : list ι), α i).
+
+variables {i : ι} {l : list ι}
+
+/-- Given `α : ι → Sort*`, a list `l` and a term `i`, as well as a term `a : α i` and a
+function `f` such that `f j : α j` for all `i'` in `m`, `pi.cons a f` is a function `g` such
+that `g i'' : α i''` for all `i''` in `i :: l`. -/
+def cons (a : α i) (f : Π j ∈ l, α j) : Π j ∈ (i :: l), α j :=
+λ j hj, if h : j = i then h.symm.rec a else f j (hj.resolve_left h)
+
+@[simp] lemma cons_eta (f : Π j ∈ (i :: l), α j) :
+  cons (f i (mem_cons_self _ _)) (λ j hj, f j (mem_cons_of_mem _ hj)) = f :=
+by { ext j hj, dsimp [cons], split_ifs with H, { cases H, refl }, { refl } }
+
+lemma map_cons (h : α i) (t : Π j ∈ l, α j)
+  {α' : ι → Sort*} (f : ∀ ⦃j⦄, α j → α' j) :
+  (λ j hj, f ((cons h t) j hj)) = cons (f h) (λ j hj, f (t j hj)) :=
+by { ext j hj, dsimp [cons], split_ifs with H, { cases H, refl }, { refl }, }
```

Comment:
More lines but fewer tactics:
```suggestion
  (λ j hj, f ((cons h t) j hj)) = cons (f h) (λ j hj, f (t j hj)) :=
begin
  ext j hj,
  refine (apply_dite (@f _) _ _ _).trans (congr_arg2 _ _ rfl),
  ext rfl,
  refl,
end
```

Is there a reason you don't state it as
```lean
  f ((cons h t) j hj) = cons (f h) (λ j hj, f (t j hj)) j hj :=
```
which looks better for forward rewrites

Reply:
In #18315 I use it for a backward rewrite. Maybe I should swap its sides.